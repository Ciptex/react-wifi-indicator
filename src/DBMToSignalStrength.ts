import type { SignalStrength } from "./SignalStrength";

/**
 * dBm is a negative number where:
 * -30 dBm is the maximum achievable signal strength. Heuristic show -42 is still an amazing, but realistic signal strength.
 * -90 dBm is unusable and approaching or drowning in the noise floor. Any functionality is highly unlikely.
 */
export const DBMToSignalStrength = (dBm: string | number): SignalStrength => {
	if ((typeof dBm !== "string" && typeof dBm !== "number") || dBm === "") {
		return "DISCONNECTED";
	}

	const num: number = Number.isFinite(dBm) ? Number(dBm) : Number(dBm);

	if (num >= -46) {
		return "EXCELLENT";
	}
	if (num >= -60) {
		return "GREAT";
	}
	if (num >= -72) {
		return "OKAY";
	}
	if (num >= -80) {
		return "WEAK";
	}
	if (num >= -90) {
		return "UNUSABLE";
	}
	return "DISCONNECTED";
};
