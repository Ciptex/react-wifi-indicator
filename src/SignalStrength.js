export default {
  'EXCELLENT': 'EXCELLENT',
  'GREAT': 'GREAT',
  'OKAY': 'OKAY',
  'WEAK': 'WEAK',
  'UNUSABLE': 'UNUSABLE',
  'DISCONNECTED': 'DISCONNECTED',
};
