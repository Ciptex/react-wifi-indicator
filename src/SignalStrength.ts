import EXCELLENT from "./images/EXCELLENT.svg";
import GREAT from "./images/GREAT.svg";
import OKAY from "./images/OKAY.svg";
import WEAK from "./images/WEAK.svg";
import UNUSABLE from "./images/UNUSABLE.svg";
import DISCONNECTED from "./images/DISCONNECTED.svg";

export type SignalStrength = "EXCELLENT" | "GREAT" | "OKAY" | "WEAK" | "UNUSABLE" | "DISCONNECTED";

export const StatusImages = {
	EXCELLENT,
	GREAT,
	OKAY,
	WEAK,
	UNUSABLE,
	DISCONNECTED
}