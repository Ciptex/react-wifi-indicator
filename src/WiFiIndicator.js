import React, { Component } from 'react';
import propTypes from './propTypes';
import defaults from './defaults';

export default class WiFiIndicator extends Component {
  static propTypes = propTypes;
  static defaultProps = defaults;

  render() {
    const { strength, statusImages, ...rest } = this.props;
    return (
      <img
        src={statusImages[strength]}
        {...rest}
      />
    )
  }
}
