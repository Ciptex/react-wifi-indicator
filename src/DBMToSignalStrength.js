import SignalStrength from './SignalStrength';

/**
 * dBm is a negative number where:
 * -30 dBm is the maximum achievable signal strength. Heuristic show -42 is still an amazing, but realistic signal strength.
 * -90 dBm is unusable and approaching or drowning in the noise floor. Any functionality is highly unlikely.
 * @param {*} dBm
 * @return {SignalStrength}
 */
const DBMToSignalStrength = (dBm) => {
  if ((typeof dBm !== 'string' && typeof dBm !== 'number') || dBm === '') {
    return SignalStrength.DISCONNECTED;
  }

  const num = Number.isFinite(dBm) ? dBm : Number(dBm);

  if (num >= -46) {
    return SignalStrength.EXCELLENT;
  }
  if (num >= -60) {
    return SignalStrength.GREAT;
  }
  if (num >= -72) {
    return SignalStrength.OKAY;
  }
  if (num >= -80) {
    return SignalStrength.WEAK;
  }
  if (num >= -90) {
    return SignalStrength.UNUSABLE;
  }
  return SignalStrength.DISCONNECTED;
};

export default DBMToSignalStrength;
