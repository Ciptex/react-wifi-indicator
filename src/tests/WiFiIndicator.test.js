import React from 'react';
import renderer from 'react-test-renderer';
import WiFiIndicator from '../WiFiIndicator';
import SignalStrength from '../SignalStrength';
import DBMToSignalStrength from '../DBMToSignalStrength';

test('snapshot #1 matches', () => {
  const tree = renderer
    .create(<WiFiIndicator strength={SignalStrength.EXCELLENT} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('snapshot #2 matches', () => {
  const tree = renderer
    .create(
      <WiFiIndicator
        strength={SignalStrength.GREAT}
        style={{ height: '240px', display: 'block' }}
        className='wifi-icon'
        alt='Awesome WiFi Signal'
      />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('snapshot #3 matches', () => {
  const tree = renderer
    .create(
      <WiFiIndicator strength={DBMToSignalStrength(-40)} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('snapshot #4 matches', () => {
  const tree = renderer
    .create(
      <WiFiIndicator
        strength={SignalStrength.WEAK}
        statusImages={{
          [SignalStrength.WEAK]: 'demo.png'
        }}
      />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

