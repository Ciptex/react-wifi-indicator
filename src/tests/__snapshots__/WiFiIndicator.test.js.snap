// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot #1 matches 1`] = `
<img
  alt=""
  src="EXCELLENT.svg"
  style={
    Object {
      "height": "3em",
    }
  }
/>
`;

exports[`snapshot #2 matches 1`] = `
<img
  alt="Awesome WiFi Signal"
  className="wifi-icon"
  src="GREAT.svg"
  style={
    Object {
      "display": "block",
      "height": "240px",
    }
  }
/>
`;

exports[`snapshot #3 matches 1`] = `
<img
  alt=""
  src="EXCELLENT.svg"
  style={
    Object {
      "height": "3em",
    }
  }
/>
`;

exports[`snapshot #4 matches 1`] = `
<img
  alt=""
  src="demo.png"
  style={
    Object {
      "height": "3em",
    }
  }
/>
`;
